### PR TITLE
Shoup's algorithm for Factorization in finite field

### PR DESCRIPTION
**Pull request by nishnik**
It is faster than Zassenhaus's when `modulo_` is large.

### CODE REVIEW

File: symengine/fields.cpp [hunk 1]
```diff
@@ -539,6 +606,100 @@ GaloisFieldDict::gf_edf_zassenhaus(const integer_class &n) const
...
+    return factors;
+}
+
+std::set<GaloisFieldDict, GaloisFieldDict::DictLess>
+GaloisFieldDict::gf_edf_shoup(const integer_class &n) const
+{
+    auto N = this->degree();
+    auto n_val = mp_get_si(n);
+    std::set<GaloisFieldDict, DictLess> factors;
+    if (N <= n_val) {
```
Comment (isuruf) [inline symengine/fields.cpp]: `N <= n`

File: symengine/fields.cpp [hunk 2]
```diff
@@ -434,6 +457,50 @@ GaloisFieldDict::gf_frobenius_map(const GaloisFieldDict &g,
     return out;
 }
 
+std::pair<GaloisFieldDict, GaloisFieldDict> GaloisFieldDict::gf_trace_map(
+    const GaloisFieldDict &a, const GaloisFieldDict &b,
+    const GaloisFieldDict &c, const integer_class &n) const
+{
+    auto n_val = mp_get_si(n);
```
Comment (isuruf) [inline symengine/fields.cpp]: Is this necessary? You can work with `n`

File: symengine/fields.cpp (hunk 1 above)
Comment (nishnik) [inline symengine/fields.cpp]: `this->degree()` is `unsigned int`

File: symengine/fields.cpp (hunk 2 above)
Comment (nishnik) [inline symengine/fields.cpp]: We have to change the value of `n`, so we have to use another variable and copy its content. I thought it would be faster on `int` than `integer_class` object, furthermore `&` operator doesn't work on `integer_class`.

File: symengine/fields.cpp (hunk 1 above)
Comment (isuruf) [inline symengine/fields.cpp]: Yes, but that doesn't mean `n` is smaller than `N` right?

Comment (nishnik) [inline symengine/fields.cpp]: I am saying that we can not compare `unsigned int` to `integer_class`

File: symengine/fields.cpp (hunk 2 above)
Comment (isuruf) [inline symengine/fields.cpp]: What is `n`? Does it fit inside a `long`? If so, why not use a `long`?

File: symengine/fields.cpp (hunk 1 above)
Comment (isuruf) [inline symengine/fields.cpp]: Why not?

Comment (nishnik) [inline symengine/fields.cpp]: I thought `integer_class` would need explicit casting into `int` for comparing, as we do in vector initialized list of `integer_class`. I will make the required changes but why is this so?

Comment (isuruf) [inline symengine/fields.cpp]: What is `n`? Does it fit into `long`? If not, then `mp_get_si(n)` will give a wrong value.

File: symengine/fields.cpp (hunk 2 above)
Comment (nishnik) [inline symengine/fields.cpp]: Yes it will fit inside `unsigned`.

Comment (isuruf) [inline symengine/fields.cpp]: Then why is a `integer_class` used?

File: symengine/fields.cpp [hunk 3]
```diff
@@ -459,9 +459,9 @@ GaloisFieldDict::gf_frobenius_map(const GaloisFieldDict &g,
 
 std::pair<GaloisFieldDict, GaloisFieldDict> GaloisFieldDict::gf_trace_map(
     const GaloisFieldDict &a, const GaloisFieldDict &b,
-    const GaloisFieldDict &c, const integer_class &n) const
+    const GaloisFieldDict &c, const unsigned long &n) const
 {
-    auto n_val = mp_get_si(n);
+    unsigned n_val(n);
```
Comment (isuruf) [inline symengine/fields.cpp]: `n` is declared `unsigned long` and `n_val` is `unsigned`. Why?

File: symengine/fields.cpp [hunk 4]
```diff
@@ -548,7 +548,7 @@ GaloisFieldDict::_gf_pow_pnm1d2(const GaloisFieldDict &f,
     return res;
 }
 
-GaloisFieldDict GaloisFieldDict::gf_random(const unsigned long &n_val,
+GaloisFieldDict GaloisFieldDict::gf_random(const unsigned &n_val,
```
Comment (isuruf) [inline symengine/fields.cpp]: Use `unsigned n_val` instead of `const unsigned &n_val`

File: symengine/fields.cpp [hunk 5]
```diff
@@ -548,7 +548,7 @@ GaloisFieldDict::_gf_pow_pnm1d2(const GaloisFieldDict &f,
...
                                            gmp_randstate_t &state) const
 {
     std::vector<integer_class> v(n_val + 1);
```
Comment (isuruf) [inline symengine/fields.cpp]: `int` should be changed to `unsigned` in line below

File: symengine/fields.cpp [hunk 6]
```diff
@@ -663,7 +663,7 @@ std::set<GaloisFieldDict, GaloisFieldDict::DictLess>
 GaloisFieldDict::gf_edf_shoup(const integer_class &n) const
 {
     auto N = this->degree();
-    auto n_val = mp_get_si(n);
+    integer_class n_val(n);
```
Comment (isuruf) [inline symengine/fields.cpp]: Why do you need a `n_val` ?

File: symengine/fields.cpp [hunk 7]
```diff
@@ -677,7 +677,7 @@ GaloisFieldDict::gf_edf_shoup(const integer_class &n) const
     auto r = gf_random(N - 1, state);
     if (modulo_ == 2_z) {
         auto h = gf_pow_mod(x, modulo_);
-        auto H = gf_trace_map(r, h, x, n - 1).second;
+        auto H = gf_trace_map(r, h, x, mp_get_si(n) - 1).second;
```
Comment (isuruf) [inline symengine/fields.cpp]: Use `mp_get_ui` if you want unsigned

File: symengine/fields.cpp [hunk 8]
```diff
@@ -663,7 +663,7 @@ std::set<GaloisFieldDict, GaloisFieldDict::DictLess>
 GaloisFieldDict::gf_edf_shoup(const integer_class &n) const
```
Comment (isuruf) [inline symengine/fields.cpp]: Is this `n` supposed to be `integer_class` or is `unsigned` enough?

File: symengine/fields.cpp (hunk 3 above)
Comment (nishnik) [inline symengine/fields.cpp]: Sorry. That's by mistake

File: symengine/fields.h [hunk 9]
```diff
@@ -229,7 +258,8 @@ class GaloisFieldDict
         GaloisFieldDict o(*this);
         for (auto &a : o.dict_) {
             a *= -1;
-            a += modulo_;
+            if (a != 0_z)
```
Comment (isuruf) [inline symengine/fields.h]: Why this change ?

File: symengine/fields.h [hunk 10]
```diff
@@ -238,7 +268,8 @@ class GaloisFieldDict
     {
         for (auto &a : dict_) {
             a *= -1;
-            a += modulo_;
+            if (a != 0_z)
+                a += modulo_;
```
Comment (isuruf) [inline symengine/fields.h]: Same here

File: symengine/fields.h [hunk 11]
```diff
@@ -285,7 +316,8 @@ class GaloisFieldDict
                 dict_.resize(other.dict_.size());
                 for (unsigned int i = orig_size; i < other.dict_.size(); i++) {
                     dict_[i] = -other.dict_[i];
-                    dict_[i] += modulo_;
+                    if (dict_[i] != 0_z)
+                        dict_[i] += modulo_;
```
Comment (isuruf) [inline symengine/fields.h]: Same here

File: symengine/fields.cpp (hunk 8 above)
Comment (nishnik) [inline symengine/fields.cpp]: We can use `unsigned`.

File: symengine/fields.h (hunk 9 above)
Comment (nishnik) [inline symengine/fields.h]: Before this change, `[0, 0, 1]` was negating to `[11, 11, 10]` in finite field(11)

File: symengine/fields.cpp [hunk 12]
```diff
@@ -289,14 +289,14 @@ bool GaloisFieldDict::gf_is_sqf() const
...
 GaloisFieldDict::gf_sqf_list() const
 {
-    std::vector<std::pair<GaloisFieldDict, integer_class>> vec_out;
+    std::vector<std::pair<GaloisFieldDict, unsigned>> vec_out;
     if (degree() < 1)
         return vec_out;
-    integer_class n = integer_class(1);
-    unsigned r = mp_get_si(modulo_);
+    unsigned n = 1;
+    unsigned r = mp_get_ui(modulo_);
```
Comment (isuruf) [inline symengine/fields.cpp]: This is unrelated to the PR, but why is this? `modulo_` can be arbitrarily large right?

Comment (nishnik) [inline symengine/fields.cpp]: It can't be, we need to have a `vector` of size `modulo_`. So there's hardware limitation.
